test/shared/src/main/scala/zio/test/mock/Spyable.scala

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test.mock
+
+import zio.{ Has, UIO, ZIO }
+
+/**
+ * `Spyable[A]` models the capability to spy on a service of type `A`.
+ * Implementations must define both a method `environment` to genereate a live
+ * environment given a mock and a method `mock` to generate a mock given a live
+ * environment. Given these definitions, it is possible to convert a service to
+ * a mock, see and modify state based on method calls, inputs, and outputs, and
+ * then convert the mock back to a live environment to spy on the service.
+ */
+trait Spyable[A] extends Mockable[A] { self =>
```

Comment:
Are there conditions in which we would want to `spy` but not `mock`? If so, we should severe the relationship between Spyable and Mockable, and make them independent.

Reply:
The issue is we need to be able to `mock` to implement `spy`. Because to `spy` we need to be able to turn a live service into a mock with `mock`, modify the mock with the specified partial function, and then turn the modified mock back into a live service with `environment` from `Mockable`.

What we could potentially do is have something like:

```scala
trait Mockable {
  def environment[A](mock: Mock): Has[A]
}

trait Unmockable {
  def mock[A](environment: Has[A]): Mock
}

trait Spyable extends Mockable with Unmockable {
  def spy = ???
}
```

That is definitely more orthogonal but it also adds more complexity so I think the question is whether the complexity is worth it, and in particular how much use you can get out of the `Unmockable` interface itself or whatever we call it. With that interface you can turn a live service into a mock and modify the mock, but you're stuck with just having a mock. So if you have a bunch of ready made `Method` values you could feed those to it and get results back, but if you wanted to interact with other systems they are going to be expecting the service being mocked and not the mock type.

Reply:
You're right, I'd leave it like it is! 👍